csharp/ql/src/semmle/code/csharp/dataflow/FrameworkDataFlow.qll

```diff
@@ -0,0 +1,502 @@
+    this = TRemoteFlowSource() and
+    result = "remote"
+    or
+    this = TLocalFlowSource() and
+    result = "local"
+    or
+    this = TStoredFlowSource() and
+    result = "stored"
+  }
+}
+
+/** Provides predicates for constructing flow-source specifications. */
+module FlowSource {
+  FlowSource remote() { result = TRemoteFlowSource() }
+
+  FlowSource local() { result = TLocalFlowSource() }
+
+  FlowSource stored() { result = TStoredFlowSource() }
+}
+
+/** A flow-sink specification. */
+class FlowSink extends TFlowSink {
+  final string toString() {
+    this = TRemoteFlowSink() and
+    result = "remote"
+    or
+    this = THtmlFlowSink() and
+    result = "html"
+    or
+    this = TEmailFlowSink() and
+    result = "email"
+    or
+    this = TSqlFlowSink() and
+    result = "sql"
+    or
+    this = TCommandFlowSink() and
+    result = "command"
+  }
+}
+
+/** Provides predicates for constructing flow-sink specifications. */
+module FlowSink {
+  FlowSink remote() { result = TRemoteFlowSink() }
+
+  FlowSink html() { result = THtmlFlowSink() }
+
+  FlowSink email() { result = TEmailFlowSink() }
+
+  FlowSink sql() { result = TSqlFlowSink() }
+
+  FlowSink command() { result = TCommandFlowSink() }
+}
+
+/** A data-flow model for a given framework. */
+abstract class FrameworkDataFlow extends string {
+  bindingset[this]
+  FrameworkDataFlow() { any() }
+
+  /**
+   * Holds if data may flow from `input` to `output` when calling `c`.
+   *
+   * `inputStack` describes the contents that is popped from the access
+   * path from the input and `outputStack` describes the contents that
+   * is pushed onto the resulting access path.
+   *
+   * `preservesValue` indicates whether this is a value-preserving step
+   * or a taint-step.
+   */
+  pragma[nomagic]
+  predicate hasSummary(
+    SourceDeclarationCallable c, SummaryInput input, ContentStack inputStack, SummaryOutput output,
+    ContentStack outputStack, boolean preservesValue
+  ) {
+    none()
+  }
+
+  /**
+   * Holds if the content stack obtained by pushing `head` onto `tail` is
+   * needed for a summary specified by `hasSummary()`.
+   *
+   * This predicate is needed for QL technical reasons only (the IPA type used
+   * to represent content stacks needs to be bounded).
+   */
+  pragma[nomagic]
+  predicate requiresContentStack(Content head, ContentStack tail) { none() }
+
+  /**
+   * Holds if values stored inside `content` are cleared on objects passed as
+   * arguments of type `input` to calls that target `c`.
+   */
+  pragma[nomagic]
+  predicate clearsContent(SourceDeclarationCallable c, SummaryInput input, Content content) {
+    none()
+  }
+
+  /** Holds if `n` is a flow source of type `source. */
+  pragma[nomagic]
+  predicate hasSource(DataFlow::Node n, FlowSource source) { none() }
+
+  /** Holds if `n` is a flow sink of type `sink. */
+  pragma[nomagic]
+  predicate hasSink(DataFlow::Node n, FlowSink sink) { none() }
+}
+
+module Examples {
```

Comment:
I would very much prefer a full example.
I.e. given some code in C#/Java/..., how does one create an appropriate extension of the `FrameworkDataFlow` class.
As an external contributor I find all this `SummaryInput`, `ContentStack`, `SummaryOutput` stuff far more complex than what I've done before (defining new data flow sinks, etc..).

Reply:
What is missing from the examples that I have already included? It shows to to define data-flow models for `StringBuilder`s, `Lazy<T>`s, and `IDbCommand`. Note that what we are modelling is the case when we are using a library that we _don't_ have the source code for.

Reply:
Well, I don't really understand what a `SummaryInput` or `ContentStack` is/does.

Reply:
The QlDoc for both classes does *not* help someone that does *not* already understand everything.
E.g. what even is a "flow-summary"?
For example:
```ql
    private predicate constructorFlow(
      Constructor c, SummaryInput input, ContentList inputAp, SummaryOutput output,
      ContentList outputAp
    ) {
      c = this.getClass().getAMember() and
      c.getParameter(0).getType() instanceof StringType and
      input = TArgumentSummaryInput(0) and
      inputAp = ContentList::empty() and
      output = TReturnSummaryOutput() and
      outputAp = ContentList::element()
    }
```

Does this mean that there is flow from an input that is the 0-th argument, to the return value of the constructor?
How does this relate to `inputAp` and `outputAp`?

Reply:
I was hoping that the QL doc for `hasSummary` was sufficient:
```ql
  /**
   * Holds if data may flow from `input` to `output` when calling `c`.
   *
   * `inputContents` describes the contents that is popped from the access
   * path from the input and `outputContents` describes the contents that
   * is pushed onto the resulting access path.
   *
   * `preservesValue` indicates whether this is a value-preserving step
   * or a taint-step.
   */
```

We can't really give examples of what `input` and `output` might be, as those are language-specific.